cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -142,28 +142,70 @@ class ExprNode extends InstructionNode {
   override string toString() { result = this.asConvertedExpr().toString() }
 }
 
+/**
+ * INTERNAL: do not use. Translates a parameter/argument index into a negative
+ * number that denotes the index of its side effect (pointer indirection).
+ */
+bindingset[index]
+int getArgumentPosOfSideEffect(int index) {
+  // -1 -> -2
+  //  0 -> -3
+  //  1 -> -4
+  // ...
+  result = -3 - index
+}
+
 /**
  * The value of a parameter at function entry, viewed as a node in a data
  * flow graph.
  */
-class ParameterNode extends InstructionNode {
-  override InitializeParameterInstruction instr;
-
+abstract class ParameterNode extends InstructionNode {
```

Comment:
I'm a bit nervous about this class becoming abstract, since it could affect customer code that extends it and I've seen that pattern used for RemoteFlowSource in other languages.

Reply:
Good point -- we don't want to expose abstract classes.

Reply:
I've pushed a commit to avoid `abstract`, make `ParameterIndirectionNode` public, and align the class names with Java's DataFlowUtil.